aea/registries/base.py

```diff
@@ -303,15 +389,18 @@ def _unregister_from_main_index(
         """
         skill_id = item_id[0]
         item_name = item_id[1]
-        name_to_item = self._items.get(skill_id, {})
-        if item_name not in name_to_item:
+        name_to_item = self._items.fetch(skill_id)
+        if name_to_item is None or item_name not in name_to_item:
             raise ValueError(
                 "No item registered with component id '{}'".format(item_id)
             )
         self.logger.debug("Unregistering item with id {}".format(item_id))
         item = name_to_item.pop(item_name)
         if len(name_to_item) == 0:
-            self._items.pop(skill_id, None)
+            self._items.unregister(skill_id)
+        else:
+            self._items.unregister(skill_id)
+            self._items.register(skill_id, name_to_item)
```

Comment:
why do we unregister and reregister?

Reply:
because the registry APIs don't have a method to `update` an item.
Also, we could rely on the side effect of the  `name_to_item` dictionary, but the APIs designed don't guarantee that strategy works.